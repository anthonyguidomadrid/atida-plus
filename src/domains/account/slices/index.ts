import { combineReducers } from '@reduxjs/toolkit'
import { createCustomerSlice } from './create-customer'
import { CreateAccountProgressSlice } from './create-account-progress'
import { loginSlice } from './login'
import { logoutSlice } from './logout'
import { setNewPasswordSlice } from './set-new-password'
import { passwordForgottenSlice } from './password-forgotten'
import { verifyResetPasswordTokenSlice } from './verify-reset-password-token'
import { customerSlice } from './customer'
import { updateCustomerPersonalDetailsSlice } from './update-customer-personal-details'
import { updateBusinessDetailsSlice } from './update-business-details'
import { updateCustomerPasswordSlice } from './update-customer-password'
import { orderHistorySlice } from './order-history'
import { orderDetailsSlice } from './order-details'
import { createNewAddressSlice } from '~domains/account/slices/create-new-address'
import { updateTaxReferenceSlice } from './update-tax-reference'
import { customerCheckSlice } from './customer-check'
import { cashBalanceSlice } from './cash-balance'
import { cashTransactionsSlice } from './cash-transactions'

export const accountReducer = combineReducers({
  'create-customer': createCustomerSlice.reducer,
  'create-account-progress': CreateAccountProgressSlice.reducer,
  login: loginSlice.reducer,
  logout: logoutSlice.reducer,
  'set-new-password': setNewPasswordSlice.reducer,
  'password-forgotten': passwordForgottenSlice.reducer,
  'verify-reset-password-token': verifyResetPasswordTokenSlice.reducer,
  customer: customerSlice.reducer,
  'update-customer-personal-details':
    updateCustomerPersonalDetailsSlice.reducer,
  'update-business-details': updateBusinessDetailsSlice.reducer,
  'create-new-address': createNewAddressSlice.reducer,
  'order-history': orderHistorySlice.reducer,
  'order-details': orderDetailsSlice.reducer,
  'update-customer-password': updateCustomerPasswordSlice.reducer,
  'update-tax-reference': updateTaxReferenceSlice.reducer,
  'customer-check': customerCheckSlice.reducer,
  'cash-balance': cashBalanceSlice.reducer,
  'cash-transactions': cashTransactionsSlice.reducer
})

export const {
  trigger: createCustomerTrigger,
  request: createCustomerRequest,
  success: createCustomerSuccess,
  failure: createCustomerFailure,
  fulfill: createCustomerFulfill,
  clear: createCustomerClear
} = createCustomerSlice.actions

export const {
  setCurrentStep,
  resetCurrentStep
} = CreateAccountProgressSlice.actions

export const {
  trigger: loginTrigger,
  triggerNotification: loginTriggerNotification,
  clearNotification: loginClearNotification,
  clearError: loginClearError,
  request: loginRequest,
  success: loginSuccess,
  failure: loginFailure,
  fulfill: loginFulfill,
  clearMessages: clearLoginMessages,
  setEmail: loginSetEmail
} = loginSlice.actions

export const {
  trigger: logoutTrigger,
  request: logoutRequest,
  success: logoutSuccess,
  failure: logoutFailure,
  fulfill: logoutFulfill
} = logoutSlice.actions

export const {
  trigger: setNewPasswordTrigger,
  request: setNewPasswordRequest,
  success: setNewPasswordSuccess,
  failure: setNewPasswordFailure,
  fulfill: setNewPasswordFulfill,
  clearState: setNewPasswordClearState
} = setNewPasswordSlice.actions

export const {
  trigger: passwordForgottenTrigger,
  request: passwordForgottenRequest,
  success: passwordForgottenSuccess,
  failure: passwordForgottenFailure,
  fulfill: passwordForgottenFulfill,
  resetState: passwordForgottenResetState
} = passwordForgottenSlice.actions

export const {
  trigger: verifyResetPasswordTokenTrigger,
  request: verifyResetPasswordTokenRequest,
  success: verifyResetPasswordTokenSuccess,
  failure: verifyResetPasswordTokenFailure,
  fulfill: verifyResetPasswordTokenFulfill
} = verifyResetPasswordTokenSlice.actions

export const {
  trigger: getCustomerTrigger,
  request: getCustomerRequest,
  success: getCustomerSuccess,
  failure: getCustomerFailure,
  fulfill: getCustomerFulfill,
  readCustomer: readCustomer,
  triggerReadCustomer: triggerReadCustomer,
  triggerShowNotification: triggerShowCustomerNotification,
  triggerHideNotification: triggerHideCustomerNotification,
  invoiceRequireTrigger: invoiceRequireTrigger,
  invoiceRequireSuccess: invoiceRequireSuccess,
  invoiceRequireFailure: invoiceRequireFailure,
  clearData: clearDataCustomer
} = customerSlice.actions

export const {
  trigger: updateCustomerPersonalDetailsTrigger,
  request: updateCustomerPersonalDetailsRequest,
  success: updateCustomerPersonalDetailsSuccess,
  failure: updateCustomerPersonalDetailsFailure,
  fulfill: updateCustomerPersonalDetailsFulfill,
  hideNotificationTrigger: updateCustomerPersonalDetailsHideNotificationTrigger,
  hideNotification: updateCustomerPersonalDetailsHideNotification
} = updateCustomerPersonalDetailsSlice.actions

export const {
  trigger: updateBusinessDetailsTrigger,
  request: updateBusinessDetailsRequest,
  success: updateBusinessDetailsSuccess,
  failure: updateBusinessDetailsFailure,
  fulfill: updateBusinessDetailsFulfill
} = updateBusinessDetailsSlice.actions

export const {
  trigger: createNewAddressTrigger,
  request: createNewAddressRequest,
  success: createNewAddressSuccess,
  failure: createNewAddressFailure,
  fulfill: createNewAddressFulfill
} = createNewAddressSlice.actions

export const {
  trigger: orderHistoryTrigger,
  request: orderHistoryRequest,
  success: orderHistorySuccess,
  failure: orderHistoryFailure,
  fulfill: orderHistoryFulfill,
  clearDetails: orderHistoryClearDetails
} = orderHistorySlice.actions

export const {
  trigger: orderDetailsTrigger,
  request: orderDetailsRequest,
  success: orderDetailsSuccess,
  failure: orderDetailsFailure,
  fulfill: orderDetailsFulfill,
  resetState: orderDetailsResetState
} = orderDetailsSlice.actions

export const {
  trigger: updateCustomerPasswordTrigger,
  request: updateCustomerPasswordRequest,
  success: updateCustomerPasswordSuccess,
  failure: updateCustomerPasswordFailure,
  fulfill: updateCustomerPasswordFulfill,
  hideNotificationTrigger: updateCustomerPasswordHideNotificationTrigger,
  hideNotification: updateCustomerPasswordHideNotification
} = updateCustomerPasswordSlice.actions

export const {
  trigger: updateTaxReferenceTrigger,
  request: updateTaxReferenceRequest,
  success: updateTaxReferenceSuccess,
  failure: updateTaxReferenceFailure,
  fulfill: updateTaxReferenceFulfill,
  clearError: clearUpdateTaxReferenceError
} = updateTaxReferenceSlice.actions

export const {
  trigger: customerCheckTrigger,
  request: customerCheckRequest,
  success: customerCheckSuccess,
  failure: customerCheckFailure,
  fulfill: customerCheckFulfill,
  clear: customerCheckClear
} = customerCheckSlice.actions

export const {
  trigger: getCashBalanceTrigger,
  request: getCashBalanceRequest,
  success: getCashBalanceSuccess,
  failure: getCashBalanceFailure,
  fulfill: getCashBalanceFulfill
} = cashBalanceSlice.actions

export const {
  trigger: getCashTransactionsTrigger,
  request: getCashTransactionsRequest,
  success: getCashTransactionsSuccess,
  failure: getCashTransactionsFailure,
  fulfill: getCashTransactionsFulfill
} = cashTransactionsSlice.actions
