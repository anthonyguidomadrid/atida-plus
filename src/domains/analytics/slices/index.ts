import { combineReducers } from '@reduxjs/toolkit'
import { isSegmentInitialisedSlice } from './is-segment-initialised'

export const analyticsReducer = combineReducers({
  data: isSegmentInitialisedSlice.reducer
})

export const { setAsInitialised } = isSegmentInitialisedSlice.actions
