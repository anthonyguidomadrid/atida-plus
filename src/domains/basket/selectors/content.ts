import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '~domains/redux'

const selectContent = (state: RootState) => state?.client?.basket?.content

export const selectIsLoading = createSelector(
  selectContent,
  content => content?.isLoading ?? false
)

export const selectPriceChannel = createSelector(
  selectContent,
  content => content.data?.priceChannel?.channel
)

export const selectBasketModalProduct = createSelector(
  selectContent,
  content => content?.basketModalProduct ?? null
)

export const selectQuantityAction = createSelector(
  selectContent,
  content => content?.quantityAction
)

export const selectWasError = createSelector(
  selectContent,
  content => content?.wasError ?? false
)

export const selectWasSuccess = createSelector(
  selectContent,
  content => content?.wasSuccess ?? false
)

export const selectError = createSelector(
  selectContent,
  content => content?.error
)

export const selectData = createSelector(
  selectContent,
  content => content?.data
)

export const selectCurrency = createSelector(selectData, data => data?.currency)

export const selectDiscountTotal = createSelector(
  selectData,
  data => data?.discountTotal
)

export const selectId = createSelector(selectData, data => data?.id)

export const selectItems = createSelector(selectData, data => data?.items ?? [])

const selectItemQuantities = createSelector(selectItems, items =>
  items.map(item => item.quantity)
)
export const selectIsProductUnavailable = createSelector(selectItems, items =>
  items.some(({ product }) => product.availability === 'NOT_AVAILABLE')
)

export const selectNumberOfItems = createSelector(
  selectItemQuantities,
  quantities =>
    quantities?.reduce((total = 0, quantity = 0) => total + quantity, 0) ?? 0
)

export const selectIsBasketEmpty = createSelector(
  selectNumberOfItems,
  numberOfItems => numberOfItems === 0
)

export const selectTotals = createSelector(selectData, data => ({
  rrpTotal: data?.rrpTotal ?? 0,
  subTotal: data?.subTotal ?? 0,
  itemTotal: data?.itemTotal ?? 0,
  shippingTotal: data?.shippingTotal,
  grandTotal: data?.grandTotal ?? 0,
  expenses: data?.expenses,
  surchargeTotal: data?.surchargeTotal,
  rrpDiscountTotal: data?.rrpDiscountTotal,
  totalSaving: data?.totalSaving ?? 0,
  rewardTotal: data?.rewardTotal,
  freeShippingThreshold: data?.freeShippingThreshold ?? 0
}))

export const selectGrandTotal = createSelector(
  selectTotals,
  totals => totals.grandTotal
)

export const selectBasketCouponData = createSelector(
  selectContent,
  content => content?.data?.coupons
)

export const selectBasketDiscounts = createSelector(selectContent, content =>
  content?.data?.discounts?.reduce((sum, item) => {
    return sum + item.amount
  }, 0)
)

export const selectCartCouponData = createSelector(
  selectContent,
  content => content?.data?.cartCoupons
)

export const selectQuantityError = createSelector(
  selectContent,
  content => content?.quantityError?.not_available_items?.[0]
)

export const selectHasPromotionalItemOutOfStock = createSelector(
  selectData,
  data => data?.hasPromotionalItemOutOfStock
)

export const selectAnonymousId = createSelector(
  selectData,
  data => data?.anonymousId
)

export const selectDiscountedItems = createSelector(
  selectData,
  data => data?.discountedItems
)

export const selectBasketDiscountsList = createSelector(
  selectData,
  data => data?.discounts
)

export const selectBasketDeliveryDays = createSelector(
  selectData,
  data => data?.deliveryDays
)
