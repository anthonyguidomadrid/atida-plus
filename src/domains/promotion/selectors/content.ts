import { createSelector } from '@reduxjs/toolkit'
import { RootState } from '~domains/redux'

const selectContent = (state: RootState) => state?.server?.promotion?.content

export const selectIsLoading = createSelector(
  selectContent,
  content => content?.isLoading ?? false
)

export const selectWasError = createSelector(
  selectContent,
  content => content?.wasError ?? false
)

export const selectWasSuccess = createSelector(
  selectContent,
  content => content?.wasSuccess ?? false
)

export const selectError = createSelector(
  selectContent,
  content => content?.error
)

export const selectData = createSelector(
  selectContent,
  content => content?.data
)

export const selectTotal = createSelector(
  selectContent,
  content => content?.total ?? 0
)

export const selectSkip = createSelector(
  selectContent,
  content => content?.skipped
)

export const selectFilters = createSelector(
  selectContent,
  content => content?.filters
)

export const selectId = createSelector(selectContent, content => content?.id)
