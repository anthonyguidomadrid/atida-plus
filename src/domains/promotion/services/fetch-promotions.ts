import {
  createClient,
  transformLocaleToUppercase,
  getCountryTagFromLocale,
  removeUndefinedPropertiesFromObject
} from '~helpers'
import { getAvailableLocales } from '~domains/translated-routes'
import {
  PromotionsQuery,
  Promotions,
  PromotionFragment
} from '~generated-graphql'
import getConfig from 'next/config'
import {
  normalizePromotion,
  normalizePromotionData,
  Promos,
  Promotion
} from '~domains/contentful/normalizers/promotion'
import {
  limitPromotions,
  skipPromotions
} from '~domains/promotion/slices/content'
import { cache } from '~helpers/server-only/redisClient'
import dayjs from 'dayjs'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'
import { FeatureFlag } from '~config/constants/feature-flags'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const fetchPromotions = async (
  locale?: string,
  skip?: string | string[],
  filters?: string[],
  preview = false
): Promise<Promos> => {
  const { serverRuntimeConfig } = getConfig()
  const shouldUseOrFormula = await loadFeatureFlag(
    locale,
    FeatureFlag.PROMOTION_OVERVIEW_PAGE_FILTERS_OR_FORMULA
  )
  const client = createClient({
    options: {
      baseURL: `${serverRuntimeConfig.cmsGraphQLHost}/spaces/${serverRuntimeConfig.cmsSpaceId}/environments/${serverRuntimeConfig.cmsEnvironmentId}`,
      headers: {
        Authorization: `Bearer ${serverRuntimeConfig.cmsToken}`
      },
      adapter: cache?.adapter
    },
    locale,
    interceptorOptions: {
      mode: 'contentful-graphql'
    }
  })

  const promotionsResponse = await client.post<{
    data: PromotionsQuery
  }>(
    '',
    removeUndefinedPropertiesFromObject({
      query: Promotions.loc?.source.body,
      variables: {
        limit: limitPromotions,
        skip: skip === undefined ? skipPromotions : parseInt(skip.toString()),
        allLocales: getAvailableLocales().map(transformLocaleToUppercase),
        locale: transformLocaleToUppercase(locale),
        where: {
          contentfulMetadata: {
            tags: {
              id_contains_all: getCountryTagFromLocale(locale)
            }
          },
          itemsToFilterBy_contains_some:
            // passing an empty array for _some results in confusing behaviour where no promos are returned!
            shouldUseOrFormula && filters && filters.length > 0
              ? filters
              : undefined,
          itemsToFilterBy_contains_all: shouldUseOrFormula
            ? undefined
            : filters,
          ...(!preview && {
            validFrom_exists: true,
            validTo_exists: true,
            validFrom_lt: dayjs().format(),
            validTo_gt: dayjs().format()
          })
        }
      }
    })
  )

  return normalizePromotionData({
    data: promotionsResponse?.data?.data?.promotionCollection ?? undefined,
    includes:
      promotionsResponse?.data?.data?.promotionCollection?.items &&
      promotionsResponse?.data?.data?.promotionCollection?.items
        .map(promotion => normalizePromotion(promotion as PromotionFragment))
        .filter((promotion): promotion is Promotion => !!promotion)
  })
}
