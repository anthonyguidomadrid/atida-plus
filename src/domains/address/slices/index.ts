import { combineReducers } from '@reduxjs/toolkit'
import { updateAddressSlice } from '~domains/address/slices/update-address'
import { createAddressSlice } from '~domains/address/slices/create-address'
import { deleteAddressSlice } from '~domains/address/slices/delete-address'
import { validateAddressSlice } from '~domains/address/slices/validate-address'

export const addressReducer = combineReducers({
  'update-address': updateAddressSlice.reducer,
  'create-address': createAddressSlice.reducer,
  'delete-address': deleteAddressSlice.reducer,
  'validate-address': validateAddressSlice.reducer
})

export const {
  trigger: createAddressTrigger,
  request: createAddressRequest,
  success: createAddressSuccess,
  failure: createAddressFailure,
  fulfill: createAddressFulfill,
  clearSuccess: createAddressClearSuccess
} = createAddressSlice.actions

export const {
  trigger: updateAddressTrigger,
  request: updateAddressRequest,
  success: updateAddressSuccess,
  failure: updateAddressFailure,
  fulfill: updateAddressFulfill,
  clearSuccess: updateAddressClearSuccess
} = updateAddressSlice.actions

export const {
  trigger: deleteAddressTrigger,
  request: deleteAddressRequest,
  success: deleteAddressSuccess,
  failure: deleteAddressFailure,
  fulfill: deleteAddressFulfill,
  itemIsLoading: deleteAddressItemIsLoading,
  itemWasSuccess: deleteAddressItemWasSuccess,
  itemWasError: deleteAddressItemWasError,
  clearSuccess: deleteAddressClearSuccess
} = deleteAddressSlice.actions

export const {
  trigger: validateAddressTrigger,
  request: validateAddressRequest,
  success: validateAddressSuccess,
  failure: validateAddressFailure,
  fulfill: validateAddressFulfill,
  reset: validateAddressReset
} = validateAddressSlice.actions
