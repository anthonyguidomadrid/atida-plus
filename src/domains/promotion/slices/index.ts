import { combineReducers } from '@reduxjs/toolkit'
import { contentSlice } from './content'

export const promotionReducer = combineReducers({
  content: contentSlice.reducer
})

export const {
  trigger: getPromotionTrigger,
  request: getPromotionRequest,
  success: getPromotionSuccess,
  failure: getPromotionFailure,
  cancel: getPromotionCancel,
  fulfill: getPromotionFulfill,
  updateSkip: handleUpdateSkip,
  reset: resetPromotion
} = contentSlice.actions
