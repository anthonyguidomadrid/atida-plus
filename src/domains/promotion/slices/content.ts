import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { clearStateErrors, hydrate } from '~domains/redux/actions'
import { Promos, Promotion } from '~domains/contentful/normalizers/promotion'
import { hasOwnProperty } from '~helpers'

export const skipPromotions = 50
export const limitPromotions = 50

type ContentState = {
  data?: Promotion[]
  isLoading: boolean
  wasSuccess: boolean
  wasError: boolean
  error?: string
  skipped?: number
  total?: number
  filters?: string[]
  id?: string
}

const initialState: ContentState = {
  data: [],
  isLoading: false,
  wasSuccess: false,
  wasError: false,
  skipped: 0,
  total: 0,
  filters: [],
  id: ''
}

export const contentSlice = createSlice({
  name: 'promotion/content',
  initialState,
  reducers: {
    trigger(
      state,
      action: PayloadAction<
        | { filters?: string[]; id?: string; skip?: string; preview?: boolean }
        | undefined
      >
    ) {
      if (action.payload?.filters && action.payload.skip === undefined) {
        // Reset data and skipped when filters have been activated: request only passes filters as param
        state.data = []
        state.skipped = 0
      }
      state.filters = action.payload?.filters ?? []
      state.id = action.payload?.id ?? ''
    },
    request(state) {
      state.isLoading = true
    },
    success(state, action: PayloadAction<Promos>) {
      state.wasSuccess = true
      state.total = action.payload.total
      action.payload.items?.forEach(item => {
        if (!state.data?.some(promo => promo.id === item.id)) {
          state.data?.push(item)
        }
      })
    },
    failure(state, action: PayloadAction<{ message: string }>) {
      const { message } = action.payload
      state.wasError = true
      state.error = message
    },
    cancel() {
      // no-op, triggered by saga when no need to re-fetch promotions
    },
    fulfill(state) {
      state.isLoading = false
    },
    reset(state) {
      state.isLoading = false
      state.wasError = false
      state.wasSuccess = false
      state.data = undefined
      state.skipped = 0
      state.total = 0
      state.filters = []
    },
    updateSkip(state, action: PayloadAction<{ newSkipValue: number }>) {
      state.wasSuccess = true
      state.skipped = action.payload.newSkipValue
    }
  },
  extraReducers: builder => {
    builder.addCase(clearStateErrors, (state, _action) => {
      state.wasError = false
      delete state.error
    })
    builder.addCase(hydrate, (_state, action) => {
      if (
        typeof action.payload === 'object' &&
        hasOwnProperty(action.payload, 'promotion') &&
        typeof action.payload.promotion === 'object' &&
        hasOwnProperty(action.payload.promotion, 'content')
      ) {
        return action.payload.promotion['content'] as ContentState
      }
    })
  }
})
