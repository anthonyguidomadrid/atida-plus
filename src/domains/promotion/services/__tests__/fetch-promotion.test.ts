/**
 * @jest-environment node
 */
import { fetchPromotions } from '../fetch-promotions'
import { contentfulPromotion } from '~components/molecules/Promotions/Promotions.mock'
// @ts-ignore
import axios from 'axios'
import { limitPromotions } from '~domains/promotion/slices/content'
import dayjs from 'dayjs'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'

describe(fetchPromotions, () => {
  describe('fetch promotions', () => {
    it('creates the client passes the locale and skip to request', async () => {
      ;(axios.post as jest.Mock).mockResolvedValue({
        data: contentfulPromotion
      })

      await fetchPromotions('en-gb', '2')
      expect(axios.post).toHaveBeenLastCalledWith('', {
        query: expect.any(String),
        variables: {
          limit: limitPromotions,
          locale: 'en-GB',
          allLocales: ['en-GB', 'pt-PT', 'es-ES', 'de-DE'],
          skip: 2,
          where: {
            contentfulMetadata: {
              tags: {
                id_contains_all: 'country-gb'
              }
            },
            itemsToFilterBy_contains_some: undefined,
            itemsToFilterBy_contains_all: undefined,
            validFrom_exists: true,
            validFrom_lt: dayjs().format(),
            validTo_exists: true,
            validTo_gt: dayjs().format()
          }
        }
      })
    })

    it('returns the normalized response', async () => {
      ;(axios.post as jest.Mock).mockResolvedValueOnce({
        data: contentfulPromotion
      })
      const response = await fetchPromotions('en-gb', '2')
      expect(response).toMatchSnapshot()
    })
  })

  describe('filtering', () => {
    describe('when OR formula is on', () => {
      beforeEach(() => {
        ;(loadFeatureFlag as jest.Mock).mockResolvedValue(true)
      })

      it('passes undefined filter when filter list is empty', async () => {
        await fetchPromotions('en-gb', '2', [])
        expect(axios.post).toHaveBeenLastCalledWith('', {
          query: expect.any(String),
          variables: {
            limit: limitPromotions,
            locale: expect.any(String),
            allLocales: expect.any(Array),
            skip: expect.any(Number),
            where: {
              contentfulMetadata: expect.any(Object),
              itemsToFilterBy_contains_some: undefined,
              itemsToFilterBy_contains_all: undefined,
              validFrom_exists: expect.any(Boolean),
              validFrom_lt: expect.any(String),
              validTo_exists: expect.any(Boolean),
              validTo_gt: expect.any(String)
            }
          }
        })
      })

      it('passes filters when filter list is not empty', async () => {
        await fetchPromotions('en-gb', '2', ['filter-one', 'filter-two'])
        expect(axios.post).toHaveBeenLastCalledWith('', {
          query: expect.any(String),
          variables: {
            limit: limitPromotions,
            locale: expect.any(String),
            allLocales: expect.any(Array),
            skip: expect.any(Number),
            where: {
              contentfulMetadata: expect.any(Object),
              itemsToFilterBy_contains_some: ['filter-one', 'filter-two'],
              itemsToFilterBy_contains_all: undefined,
              validFrom_exists: expect.any(Boolean),
              validFrom_lt: expect.any(String),
              validTo_exists: expect.any(Boolean),
              validTo_gt: expect.any(String)
            }
          }
        })
      })
    })

    describe('when OR formula is off', () => {
      beforeEach(() => {
        ;(loadFeatureFlag as jest.Mock).mockResolvedValue(false)
      })

      it('passes undefined filter when filter list is empty', async () => {
        await fetchPromotions('en-gb', '2', [])
        expect(axios.post).toHaveBeenLastCalledWith('', {
          query: expect.any(String),
          variables: {
            limit: limitPromotions,
            locale: expect.any(String),
            allLocales: expect.any(Array),
            skip: expect.any(Number),
            where: {
              contentfulMetadata: expect.any(Object),
              itemsToFilterBy_contains_some: undefined,
              itemsToFilterBy_contains_all: [],
              validFrom_exists: expect.any(Boolean),
              validFrom_lt: expect.any(String),
              validTo_exists: expect.any(Boolean),
              validTo_gt: expect.any(String)
            }
          }
        })
      })

      it('passes filters when filter list is not empty', async () => {
        await fetchPromotions('en-gb', '2', ['filter-one', 'filter-two'])
        expect(axios.post).toHaveBeenLastCalledWith('', {
          query: expect.any(String),
          variables: {
            limit: limitPromotions,
            locale: expect.any(String),
            allLocales: expect.any(Array),
            skip: expect.any(Number),
            where: {
              contentfulMetadata: expect.any(Object),
              itemsToFilterBy_contains_some: undefined,
              itemsToFilterBy_contains_all: ['filter-one', 'filter-two'],
              validFrom_exists: expect.any(Boolean),
              validFrom_lt: expect.any(String),
              validTo_exists: expect.any(Boolean),
              validTo_gt: expect.any(String)
            }
          }
        })
      })
    })
  })
})
