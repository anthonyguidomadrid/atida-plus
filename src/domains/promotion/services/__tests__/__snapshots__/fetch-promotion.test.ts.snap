// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPromotions fetch promotions returns the normalized response 1`] = `
Object {
  "items": Array [
    Object {
      "categories": Array [
        Object {
          "id": "beauty",
          "path": Object {
            "id0": "beauty",
          },
        },
        Object {
          "id": "beauty_body",
          "path": Object {
            "id0": "beauty_body",
          },
        },
      ],
      "color": "category-personal-care",
      "contentType": "Promotion",
      "description": "<p>This is a test promotion 50% of Tropicania</p>",
      "id": "marjolein.test",
      "image": Object {
        "url": "https://images.ctfassets.net/7g2w796onies/2FxL8j1drC2fIORm0MPbas/d151066571b4b7d4e8024909abf54afb/AVene.png",
      },
      "isContentWithImage": true,
      "isSponsoredContent": false,
      "labels": Array [
        Object {
          "label": "campaign-promo.black-friday.promo.2021",
          "type": "campaign-promo",
        },
        Object {
          "label": "Whoop whoop 2",
          "type": "promotion",
        },
      ],
      "promoInformation": "",
      "teaserDescription": "Whoop whoop",
      "title": "Marjolein Test promo",
      "url": "/marjolein-test-promo",
      "validFrom": "2021-06-04T00:00:00.000+02:00",
      "validTo": "2021-06-11T00:00:00.000+02:00",
    },
  ],
  "total": 8,
}
`;
